src/types.rs

```diff
@@ -11,3 +14,32 @@ impl Mode {
         self.0
     }
 }
+
+#[repr(transparent)]
+pub struct CStr(str);
+
+impl CStr {
+    pub fn new<'a>(data: &'a str) -> &'a CStr {
+        if data.bytes().position(|b| b == b'\x00') != Some(data.len() - 1) {
+            panic!("CStr must contain a single NUL byte at the end");
+        }
+        unsafe {
+            return mem::transmute(data);
```

Comment:
Why does this need a `transmute` - why can't you return `CStr(*data)` or something?

Reply:
Because this function returns a reference, and `CStr` constructs a new value, so you can't return a reference because then it just points to something on the stack. DST are magic.